Add integration tests for elasticsearch #2244

lordlothar99 opened the pull request.

deepu105 commented on generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java:
```diff
@@ -4,8 +4,8 @@
 import <%=packageName%>.<%= mainClass %>;
 import <%=packageName%>.domain.<%= entityClass %>;
 import <%=packageName%>.repository.<%= entityClass %>Repository;<% if (service != 'no') { %>
-import <%=packageName%>.service.<%= entityClass %>Service;<% } else { if (searchEngine == 'elasticsearch') { %>
```
why did you remove the else statement?  when there is service layer tests should be triggered via that instead of repository

lordlothar99 replied:
Direct access to SearchRepository is needed whether service is present or not, cause some new assertions are added in create / update / delete test methods

deepu105 replied:
ok since you are using `<%= entityInstance %>SearchRepository.deleteAll();` anyway

deepu105 commented on generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java:
```diff
@@ -195,7 +195,8 @@ public void setup() {
     public void initTest() {<% if (databaseType == 'mongodb' || databaseType == 'cassandra') { %>
-        <%= entityInstance %>Repository.deleteAll();<% } %>
+        <%= entityInstance %>Repository.deleteAll();<% } if (searchEngine == 'elasticsearch') { %>
+        <%= entityInstance %>SearchRepository.deleteAll();<% } %>
```
shouldnt this be handled by the ` <%= entityInstance %>Repository.deleteAll();`

deepu105 replied:
Ok this is needed for non service entity, if entity has service this is handled in its delete method

lordlothar99 replied:
yes, we need both `deleteAll()` invocations

geraldhumphries replied:
If the mapping of an entity changes, trying to insert an object into an old index can cause exceptions (ex. if a field type changes). Deleting the data is not enough, you need to delete the entity's index then recreate it with the new mapping:
```
        elasticsearchTemplate.deleteIndex(<%= entityClass %>.class);
        elasticsearchTemplate.createIndex(<%= entityClass %>.class);
        elasticsearchTemplate.putMapping(<%= entityClass %>.class);
```

This will also delete the data.

lordlothar99 replied:
yeah I read that, actually I was wondering we could just use `deleteAll()`, as it's fast, and during the tests mappings won't change .. no?

geraldhumphries replied:
They won't change during the tests, but the ES indexes persist (they're stored on the hard drive) so the mappings may change between test sessions. That's why I was trying to find a solution to delete all indexes before any test is run.

deepu105 replied:
Ill look at them today
On 17 Mar 2016 03:38, "Gerald Humphries" <notifications@github.com> wrote:

> In
> generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java
> <https://github.com/jhipster/generator-jhipster/pull/3168#discussion_r56400756>
> :
>
> > @@ -195,7 +195,8 @@ public void setup() {
> >
> >      @Before
> >      public void initTest() {<% if (databaseType == 'mongodb' || databaseType == 'cassandra') { %>
> > -        <%= entityInstance %>Repository.deleteAll();<% } %>
> > +        <%= entityInstance %>Repository.deleteAll();<% } if (searchEngine == 'elasticsearch') { %>
> > +        <%= entityInstance %>SearchRepository.deleteAll();<% } %>
>
> They won't change during the tests, but the ES indexes persist (they're
> stored on the hard drive) so the mappings may change between test sessions.
> That's why I was trying to find a solution to delete all indexes before any
> test is run.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/jhipster/generator-jhipster/pull/3168/files/5f9abafc5f318de35a7b91a1c4196c26afaa7446#r56400756>
>

deepu105 replied:
Ill let you guys decide this

lordlothar99 replied:
@geraldhumphries ok, I understand now that I've read once again your comment. I will try to find a better way to handle that...

deepu105 commented on generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java:
```diff
@@ -231,7 +232,12 @@ public void initTest() {<% if (databaseType == 'mongodb' || databaseType == 'cas
+        <%_ }} if (searchEngine == 'elasticsearch') { _%>
+
+        // Validate the <%= entityClass %> in ElasticSearch
+        <%= entityClass %> <%= entityInstance %>Es = <%= entityInstance %>SearchRepository.findOne(test<%= entityClass %>.getId());
```
this should be `<%= entityClass %>Service.findOne` when entity has service and else `<%= entityInstance %>SearchRepository.findOne`

lordlothar99 replied:
Actually , `EntityService.findOne()` invokes `Entity repository.findOne()`.. While the purpose of this assertion is to read from Elasticsearch. So I need direct usage of `EntitySearchRepository.findOne()`

geraldhumphries replied:
What about sending a request to `_search/entities` with the query `"id:" + test<%= entityClass %>.getId()`?

geraldhumphries replied:
Never mind, I see that it is following the same pattern as above.

deepu105 replied:
ok

deepu105 commented on generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java:
```diff
@@ -308,7 +314,8 @@ public void initTest() {<% if (databaseType == 'mongodb' || databaseType == 'cas
         // Initialize the database
-        <%= entityInstance %>Repository.save<% if (databaseType == 'sql') { %>AndFlush<% } %>(<%= entityInstance %>);
+        <%= entityInstance %>Repository.save<% if (databaseType == 'sql') { %>AndFlush<% } %>(<%= entityInstance %>);<% if (searchEngine == 'elasticsearch') { %>
+        <%= entityInstance %>SearchRepository.save(<%= entityInstance %>);<%_ } _%>
```
this is required only if entity doesnt have a service else its handled in the service already

lordlothar99 replied:
You are right. I just wanted to keep the code similar to the other tests.. But I can change that if you prefer service invocation?

deepu105 replied:
yes since service already does that its a good way to test if service does that so i would prefer to use service when its enabled

lordlothar99 replied:
ok I will

deepu105 commented:
can you take care of my comments plz. once done ill merge it

lordlothar99 commented:
Hi @deepu105, just let me know what u think about my explanations plz

deepu105 commented:
Im ok with all except the last save `<%= entityInstance %>SearchRepository.save(<%= entityInstance %>);` stuff I would prefer that via service when its enabled

lordlothar99 commented:
I'm on it

lordlothar99 commented:
@deepu105 as we said, I replaced `EntityRepository.save()` and `EntitySearchRepository.save()` by `EntityService.save()` when possible
@geraldhumphries I introduced new `EntityIndexInitializer` beans, which reset ElasticSearch indexes at startup only (`@PostConstruct` method), it takes more or less 2s on my (slow) computer. Still `EntityResourceIntTest` `@Before` method invokes `EntitySearchRepository.deleteAll()` which is very fast (<10ms).
Can you please review these changes, and let me know ?
thx !

lordlothar99 commented:
seems like Travis is failing sometimes, but I think that's nothing to do with this PR, no ?

jdubois commented:
No we are having other build issues at the moment, don't worry!
Le 18 mars 2016 9:11 AM, "François Lecomte" <notifications@github.com> a
écrit :

> seems like Travis is failing sometimes, but I think that's nothing to do
> with this PR, no ?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/jhipster/generator-jhipster/pull/3168#issuecomment-198246808>
>

cicoub13 commented on generators/entity/templates/src/test/java/package/web/rest/_EntityResourceIntTest.java:
```diff
@@ -361,6 +382,35 @@ public void initTest() {<% if (databaseType == 'mongodb' || databaseType == 'cas
+        assertThat(<%= entityInstancePlural %>).hasSize(databaseSizeBeforeDelete - 1);<% if (searchEngine == 'elasticsearch') { %>
+
+        // Validate tElasticSearch is empty
+        boolean <%= entityInstance %>ExistsInEs = <%= entityInstance %>SearchRepository.exists(<%= entityInstance %>.getId());
```
When entityInstance is same that entityInstancePlural, there is a problem here (with entity people for example). 

`<%= entityInstance %>.getId()` does not exist because entityInstance was redefined in a `List<<%= entityClass %>>`

deepu105 replied:
@lordlothar99 can you check this plz?

cicoub13 replied:
Inversing elasticSearch validation and database validation works.
Replcaing `List<<%= entityClass %>> <%= entityInstancePlural %>` by `List<<%= entityClass %>> <%= entityInstancePlural %>Database` works too.

lordlothar99 replied:
I'm on it ; do you prefer a new PR, or update this one ?

deepu105 replied:
I think you cant update this as its merged you need to do a new PR

lordlothar99 replied:
yeah u r right, I pushed but nothing happened ...
I created fixed #3227
